cmd/testbeacon.go

```diff
@@ -0,0 +1,148 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package cmd
+
+import (
+	"context"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/maps"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+type testBeaconConfig struct {
+	testConfig
+	Endpoints []string
+}
+
+func newTestBeaconCmd(runFunc func(context.Context, io.Writer, testBeaconConfig) error) *cobra.Command {
+	var config testBeaconConfig
+
+	cmd := &cobra.Command{
+		Use:   "beacon",
+		Short: "Run multiple tests towards beacon nodes",
+		Long:  `Run multiple tests towards beacon nodes. Verify that Charon can efficiently interact with Beacon Node(s).`,
+		Args:  cobra.NoArgs,
+		PreRunE: func(cmd *cobra.Command, _ []string) error {
+			return mustOutputToFileOnQuiet(cmd)
+		},
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return runFunc(cmd.Context(), cmd.OutOrStdout(), config)
+		},
+	}
+
+	bindTestFlags(cmd, &config.testConfig)
+	bindTestBeaconFlags(cmd, &config)
+
+	return cmd
+}
+
+func bindTestBeaconFlags(cmd *cobra.Command, config *testBeaconConfig) {
+	const endpoints = "endpoints"
+	cmd.Flags().StringSliceVar(&config.Endpoints, endpoints, nil, "[REQUIRED] Comma separated list of one or more beacon node endpoint URLs.")
+	mustMarkFlagRequired(cmd, endpoints)
+}
+
+func supportedBeaconTestCases() map[testCaseName]func(context.Context, *testBeaconConfig) testResult {
+	return map[testCaseName]func(context.Context, *testBeaconConfig) testResult{
+		{name: "ping", order: 1}: beaconPing,
+	}
+}
+
+func runTestBeacon(ctx context.Context, w io.Writer, cfg testBeaconConfig) (err error) {
+	testCases := supportedBeaconTestCases()
+	supportedTestCases := maps.Keys(testCases)
+
+	queuedTests, err := filterTests(supportedTestCases, cfg.testConfig)
+	if err != nil {
+		return err
+	}
+	sort.Slice(queuedTests, func(i, j int) bool {
+		return queuedTests[i].order < queuedTests[j].order
+	})
+
+	parentCtx := ctx
+	if parentCtx == nil {
+		parentCtx = context.Background()
+	}
+	timeoutCtx, cancel := context.WithTimeout(parentCtx, cfg.Timeout)
+	defer cancel()
+
+	ch := make(chan testResult)
+	res := testCategoryResult{
+		TestsExecuted: make(map[string]testResult),
+		CategoryName:  "beacon",
+	}
+
+	startTime := time.Now()
+	// run all beacon tests, pushing each finished test until all are finished or timeout occurs
+	go runAllBeacon(timeoutCtx, queuedTests, testCases, cfg, ch)
+
+	testCounter := 0
+outer:
+	for {
+		var name string
+		select {
+		case <-timeoutCtx.Done():
+			name = queuedTests[testCounter].name
+			res.TestsExecuted[name] = testResult{Verdict: testVerdictFail}
+
+			break outer
+		case result, ok := <-ch:
+			if !ok {
+				break outer
+			}
+			name = queuedTests[testCounter].name
+			testCounter++
+			res.TestsExecuted[name] = result
+		}
+	}
+
+	res.ExecutionTime = Duration{time.Since(startTime)}
+	res.Score = calculateScore(res.TestsExecuted)
```

Comment:
I've been thinking about that and thought - what use can a user of ours make of the measurements, without our score attached to it? I think almost all of the users wouldn't know if 100ms is good or bad for a specific call? If we don't tell them what is sufficient and what is not, they can't really make up a conclusion on their own?

Reply:
Thing is, an user might want to upload their measurements to a Launchpad page to e.g. prove to the other peers that their node is up for the task. The launchpad might have a scoring system embedded in it, and can show a pretty UI around it.

Reply:
Ah, okay, got what you mean. But then if we offload this logic to the launchpad, we are coupling the scoring with the launchpad, should those two be related? For me it has more connection to the tests in charon. We can still have the same functionality in the launchpad, it will just consume the already evaluated scores. Then also the non-launchpad users will benefit from the scoring (if I do the whole flow through the CLI, I won't have to go to the launchpad to get scores).

Reply:
I generally agree about making it as universal as possible, but maybe let's keep it as simple as possible at this stage of design/development? Just to move forward quicker. I believe this code is not very complex to not change it in the near future. I suspect we will change it anyway as soon as we actually implement a few tests and people start using them.

Reply:
LFTM!